### PR TITLE
DSC repository (EXPOSUREAPP-8363)

### PR DESCRIPTION
**Pull request by jurajkusnier**
DSC repository
- load data from local storage
- load data from hardcoded asset
- load data from API

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParser.kt [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+import org.joda.time.Instant
 import javax.inject.Inject
 
 @Reusable
 class DscDataParser @Inject constructor() {
 
-    fun parse(rawData: ByteArray): DscData {
-        throw NotImplementedError()
-    }
+    fun parse(rawData: ByteArray, updatedAt: Instant = Instant.EPOCH): DscData = DscData(
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParser.kt:12]: can we  make default value `Instant.now()` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
 @Singleton
 class DefaultDscSource @Inject constructor(
+    @AppContext private val context: Context,
     private val dscDataParser: DscDataParser,
 ) {
 
     fun getDscData(): DscData {
-        Timber.d("getDscData()")
-        throw NotImplementedError()
+        val rawData = context.assets.open("default_dsc_list.bin").readBytes()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt:17]:
```kotlin 
context.assets.open("default_dsc_list.bin").use { it.readBytes() }  ?
```

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParserTest.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
...1UEBhMCREUxCzAJBgNVBAgMAkJXMREwDwYDVQQHDAhXYWxsZG9yZjEPMA0GA1UECgwGU0FQIFNFMRAwDgYDVQQLDAdDV0EgQ0xJMRAwDgYDVQQDDAdjd2EtY2xpMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4AX+nC5rHQd8VZbsATS+jO7LtLNZUs/Lo/YMw1/tb5si1xNHJel0DqiQj0rYf8ylPWq+4ghPcOrZr62e9yVwDaNSMFAwDgYDVR0PAQH/BAQDAgeAMB0GA1UdDgQWBBRpZH0bn2jUDNGU344Zi0gYSxBTTTAfBgNVHSMEGDAWgBS/hghB7Sqx/Z49r2OWQiJCveyCOTAJBgcqhkjOPQQBA0kAMEYCIQCro3URao1c+tnDrlII5bWQajXZl8bPKTDOZfG5IpAvfwIhAMv146/BKNIubPTJd6/PDZ+gyKiNp4j+UURF4AhD+EC9Cp4ECgiVVSD8UxJFXhKRBDCCAg0wggGzoAMCAQICCQDRh2ekBKNX+jAKBggqhkjOPQQDAjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwHhcNMzEwMTA0MTIwNTUyWhcNNDEwMTAxMTIwNTUyWjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAARfGw+Ed6QubOCEoglZGK6IP12ECbDCgw2ghO8IVtfFuFjEgVDe5pSP7OdShpqlIAr874EVJDv4hrp5vu/IfYG7o1IwUDAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0OBBYEFFOV0esQ3mWGV0QA38qrb5+zORn7MB8GA1UdIwQYMBaAFL+GCEHtKrH9nj2vY5ZCIkK97II5MAoGCCqGSM49BAMCA0gAMEUCIB5sc2e5+AIpchjkFnUO+6sW6eojNO7NASC7CQEvq71zAiEAoUusMyFeIgjhwubhgcGbpOzAso0uYuD0iycljd81nlo="
```
Comment (mtwalli) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParserTest.kt:33]: should we use the same text in `DscSignatureValidator` ?

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServerTest.kt [hunk 4]
```diff
@@ -74,5 +74,8 @@ internal class DscServerTest {
...
...1UEBhMCREUxCzAJBgNVBAgMAkJXMREwDwYDVQQHDAhXYWxsZG9yZjEPMA0GA1UECgwGU0FQIFNFMRAwDgYDVQQLDAdDV0EgQ0xJMRAwDgYDVQQDDAdjd2EtY2xpMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4AX+nC5rHQd8VZbsATS+jO7LtLNZUs/Lo/YMw1/tb5si1xNHJel0DqiQj0rYf8ylPWq+4ghPcOrZr62e9yVwDaNSMFAwDgYDVR0PAQH/BAQDAgeAMB0GA1UdDgQWBBRpZH0bn2jUDNGU344Zi0gYSxBTTTAfBgNVHSMEGDAWgBS/hghB7Sqx/Z49r2OWQiJCveyCOTAJBgcqhkjOPQQBA0kAMEYCIQCro3URao1c+tnDrlII5bWQajXZl8bPKTDOZfG5IpAvfwIhAMv146/BKNIubPTJd6/PDZ+gyKiNp4j+UURF4AhD+EC9Cp4ECgiVVSD8UxJFXhKRBDCCAg0wggGzoAMCAQICCQDRh2ekBKNX+jAKBggqhkjOPQQDAjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwHhcNMzEwMTA0MTIwNTUyWhcNNDEwMTAxMTIwNTUyWjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAARfGw+Ed6QubOCEoglZGK6IP12ECbDCgw2ghO8IVtfFuFjEgVDe5pSP7OdShpqlIAr874EVJDv4hrp5vu/IfYG7o1IwUDAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0OBBYEFFOV0esQ3mWGV0QA38qrb5+zORn7MB8GA1UdIwQYMBaAFL+GCEHtKrH9nj2vY5ZCIkK97II5MAoGCCqGSM49BAMCA0gAMEUCIB5sc2e5+AIpchjkFnUO+6sW6eojNO7NASC7CQEvq71zAiEAoUusMyFeIgjhwubhgcGbpOzAso0uYuD0iycljd81nlo="
```
Comment (mtwalli) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServerTest.kt:79]: I think it is better if extract it and use it everywhere

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorageTest.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
...1UEBhMCREUxCzAJBgNVBAgMAkJXMREwDwYDVQQHDAhXYWxsZG9yZjEPMA0GA1UECgwGU0FQIFNFMRAwDgYDVQQLDAdDV0EgQ0xJMRAwDgYDVQQDDAdjd2EtY2xpMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4AX+nC5rHQd8VZbsATS+jO7LtLNZUs/Lo/YMw1/tb5si1xNHJel0DqiQj0rYf8ylPWq+4ghPcOrZr62e9yVwDaNSMFAwDgYDVR0PAQH/BAQDAgeAMB0GA1UdDgQWBBRpZH0bn2jUDNGU344Zi0gYSxBTTTAfBgNVHSMEGDAWgBS/hghB7Sqx/Z49r2OWQiJCveyCOTAJBgcqhkjOPQQBA0kAMEYCIQCro3URao1c+tnDrlII5bWQajXZl8bPKTDOZfG5IpAvfwIhAMv146/BKNIubPTJd6/PDZ+gyKiNp4j+UURF4AhD+EC9Cp4ECgiVVSD8UxJFXhKRBDCCAg0wggGzoAMCAQICCQDRh2ekBKNX+jAKBggqhkjOPQQDAjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwHhcNMzEwMTA0MTIwNTUyWhcNNDEwMTAxMTIwNTUyWjBiMQswCQYDVQQGEwJERTELMAkGA1UECAwCQlcxETAPBgNVBAcMCFdhbGxkb3JmMQ8wDQYDVQQKDAZTQVAgU0UxEDAOBgNVBAsMB0NXQSBDTEkxEDAOBgNVBAMMB2N3YS1jbGkwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAARfGw+Ed6QubOCEoglZGK6IP12ECbDCgw2ghO8IVtfFuFjEgVDe5pSP7OdShpqlIAr874EVJDv4hrp5vu/IfYG7o1IwUDAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0OBBYEFFOV0esQ3mWGV0QA38qrb5+zORn7MB8GA1UdIwQYMBaAFL+GCEHtKrH9nj2vY5ZCIkK97II5MAoGCCqGSM49BAMCA0gAMEUCIB5sc2e5+AIpchjkFnUO+6sW6eojNO7NASC7CQEvq71zAiEAoUusMyFeIgjhwubhgcGbpOzAso0uYuD0iycljd81nlo="
```
Comment (mtwalli) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorageTest.kt:64]: here too

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt [hunk 6]
```diff
@@ -31,17 +32,24 @@ class DscRepository @Inject constructor(
...
     val dscData = internalData.data
 
     suspend fun refresh() {
         Timber.tag(TAG).d("refresh()")
+        internalData.updateBlocking {
+            dscServer.getDscList().let { rawData ->
+                mapDscList(rawData).apply {
+                    localStorage.save(rawData)
+                }
+            }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt:47]: what happens if `getDscList` or `mapDscList` throw and error ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt [hunk 7]
```diff
@@ -1,26 +1,57 @@
...
 
-    suspend fun save(rawData: ByteArray?): Unit = mutex.withLock {
-        Timber.d("save(rawData=${rawData?.size}")
-        throw NotImplementedError()
+    suspend fun save(rawData: ByteArray): Unit = mutex.withLock {
+        Timber.v("save (rawData.size=${rawData.size}")
+
+        if (dscDir.mkdirs()) Timber.v("Parent folder created.")
+
+        if (dscFile.exists()) {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt:47]: where the file is created ? I don't see `createNewFile`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParser.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParser.kt]: we can, I don't have preferences

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt]: updated

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParserTest.kt (hunk 3 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscDataParserTest.kt]: updated

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServerTest.kt (hunk 4 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/test/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServerTest.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt (hunk 7 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt:47]: `dscFile.writeBytes(rawData)` create the file if it doesn't exist or overwrite it if already exists. 
See `LocalDscStorageTest`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt (hunk 6 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt:46]: It's handled by `DccStateCheckScheduler`
```kotlin
            try {
                    Timber.tag(TAG).i("Refreshing DSC data.")
                    dscRepository.refresh()
                } catch (e: Exception) {
                    Timber.tag(TAG).e(e, "Refreshing DSC data failed.")
                }
```

**Comment by mtwalli** [comment]
Should we delete the local data when environment changes ? and on app reset?

when App opens on INT and env changed it keeps reading from local saved INT

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt [hunk 8]
```diff
@@ -32,18 +31,24 @@ class DscRepository @Inject constructor(
             replayExpirationMillis = 0
         ),
     ) {
-
-        // TODO
-        DscData(
-            dscList = emptyList(),
-            updatedAt = Instant.EPOCH,
-        )
+        localStorage.load() ?: defaultDscData.getDscData()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscRepository.kt:34]: can you please add some log here ? to help spot what is going on?
maybe 
```kotlin
        Timber.tag(TAG).d("internalData")

        val localData = localStorage.load()
        if (localData != null) {
            Timber.tag(TAG).d("localStorage data exists")
            localData
        } else {
            Timber.tag(TAG).d("Reading from defaultDscData")
            defaultDscData.getDscData()
        }
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServer.kt [hunk 9]
```diff
@@ -20,7 +19,7 @@ class DscServer @Inject constructor(
     private val dscApi: DscApiV1
 ) {
 
-    suspend fun getDscList(): DscList {
+    suspend fun getDscList(): ByteArray {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/server/DscServer.kt:22]: and here ?
```
   Timber.tag(TAG).d("getDscList()")
```

**Comment by jurajkusnier** [comment]
> Should we delete the local data when environment changes ? and on app reset?
> 
> when App opens on INT and env changed it keeps reading from local saved INT

I didn't find this requirement in the specs. But It's actually not personal data so I don't see a reason for removing it and falling back to the hardcoded list. Regarding env change, we never remove anything as it's just a feature for debugging

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt [hunk 10]
```diff
@@ -1,18 +1,21 @@
...
 import de.rki.coronawarnapp.covidcertificate.signature.core.DscDataParser
-import timber.log.Timber
+import de.rki.coronawarnapp.util.di.AppContext
+import org.joda.time.Instant
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class DefaultDscSource @Inject constructor(
+    @AppContext private val context: Context,
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/DefaultDscSource.kt:13]: yes

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt [hunk 11]
```diff
@@ -1,26 +1,57 @@
...
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class LocalDscStorage @Inject constructor(
+    @AppContext context: Context,
     private val dscDataParser: DscDataParser,
 ) {
     private val mutex = Mutex()
+    private val dscDir = File(context.filesDir, "dsc_storage")
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt:21]: That's how we structured the code in the beginning. It's the same implementation as we are using for AppConfig or other data downloaded from CDN.
We have default data source (hardcoded asset file), local data source (latest response from backed saved into filesDir), and remote data source (API)

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt [hunk 12]
```diff
@@ -1,26 +1,57 @@
...
+    private val dscFile = File(dscDir, "dsclist")
 
-    suspend fun load(): DscData = mutex.withLock {
-        Timber.d("load()")
-        throw NotImplementedError()
+    suspend fun load(): DscData? = mutex.withLock {
+        Timber.v("load()")
+
+        if (dscFile.exists()) {
```
[the remark this answers is not in the document]
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt:27]: I rather don't use early return, but I guess it's just a matter of opinion, both have pros and cons 🤷

**Review by mtwalli — APPROVED** [review]
Lgtm, tested against mock server

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (d4rken) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/storage/LocalDscStorage.kt:21]: Hm, you are right, makes more sense to have it not `cache`. It's part of the QR Code scan flow which can also be done offline.

**Review by d4rken — APPROVED** [review]

**Pull request merged**